spec/index.md

```diff
@@ -2095,6 +2095,24 @@ If you do not specify the `max_response_bytes` parameter, the maximum of a `2MB`
 
 :::
 
+### IC method `node_metrics_history` {#ic-node-metrics-history}
+
+:::note
+
+The node metrics management canister API is considered EXPERIMENTAL. Canister developers must be aware that the API may evolve in a non-backward-compatible way.
+
+:::
+
+Given a subnet ID as input, this method returns collections of metrics for all nodes on the given subnet (field `node_metrics`) at timestamps provided in nanoseconds since 1970-01-01 (field `timestamp_nanos`) at which the metrics were sampled. The returned timestamps are all timestamps after (and including) the provided timestamp (field `start_at_timestamp_nanos`) for which node metrics are available. The maximum number of returned timestamps is 60 and no two returned timestamps belong to the same UTC day.
```

Comment:
This requirement that "no two returned timestamps belong to the same UTC day" makes the actual timestamps quite unwieldy.

I.e. ideally one would expect one sample per day, recorded at midnight UTC. But that would mean that we could never return the "current" metrics (because there would be a sample at "midnight today" and the spec does not allow for another sample for "now"). And I would argue that the ability to query the current stats (at arbitrary intervals) is at least as important as the ability to query historical stats.

As is, the implementation is forced to record samples with timestamps just before midnight UTC. Whether that's e.g. always one second before midnight; or (as currently implemented) the actual timestamp of the last observation before midnight (which introduces arbitrary jitter).

Reply:
I hear you @alin-at-dfinity , and I agree with you - if I understood correctly your concern.
Would you mind going ahead and proposing some changes that would improve the interface? Would the change be as simple as removing this limitation `no two returned timestamps belong to the same UTC day` and thus allowing two entries for the current day: one for period since 0:00 UTC and another from now-delta?

Reply:
Currently the implementation returns an iterator over all snapshots, including the current metrics as the last element. I guess it would be easy to change if we want it differently, we just need to decide.

I also wonder whether we should say something about counter resets: if a node doesn't report metrics for one day before a snapshot the implementation will not record its counter in this snapshot; similar to how Prometheus handles this. It seems to be important to mention on the interface level so that users end up with the right values when computing differences between several days.

Reply:
Following up on @yvonneanne's comment here to stick to one thread.

> Resets: Afaiu, no resets happen, the counters can only go up, or am I
> missing something?

Not sure I understand. If a node is gone for some time (the implementation currently uses 1 day) then it is garbage collected to bound the size of the recorded metrics. When it re-appears its counter will start from 0. So for the raw metrics in the replicated state it is not always guaranteed that they only go up. We could also offer an API function which hides this detail from the user, but we'd currently prefer not to because it keeps the replica implementation simple and it is consistent with the way Prometheus handles such things. 

Or are you saying that it will never happen that a node disappears for a day?

> Last element: To keep the interface cleaner I'd follow Alin's suggestion to
> have a separate endpoint for this to avoid any confusion and ambiguity.

No strong opinion. Once there is consensus please ping us so that we can provide an appropriate interface.

Reply:
@yvonneanne
> Resets: Afaiu, no resets happen, the counters can only go up, or am I missing something?

Resets will happen (as currently implemented) if a node is removed from the subnet and later added back.

We probably shouldn't hold on to stats for nodes have long since been removed from the subnet. So we have 2 options:
* we drop node stats as soon as they stop being updated (i.e. the day after the node has been removed); or
* we keep node stats around for 60 days (or however long the sliding window is) and (as soon as the stats are identical across all snapshots) go back and remove the node from all snapshots.

The latter has the downside that we're changing history (the stats from one week ago may be different depending on when they were queried), so we went with the first option (which also simplifies the implementation).

However, if the consensus is to add a separate Management Canister method for current metrics (as opposed to all snapshots after some timestamp); then without changing the implementation we could artificially populate removed nodes' stats  in the response to `node_metrics_history`.Or we may return the raw stats, something that may result in counter resets. Or at least in some node's stats being present in the first snapshot, but not in the last.

Regardless of which approach we choose, it may be worth specifying it explicitly.

Reply:
> Regardless of which approach we choose, it may be worth specifying it explicitly.

+1

Reply:
> Following up on @yvonneanne's comment here to stick to one thread.
Thanks, David (I used the email reply button, clearly this was a mistake)

> Not sure I understand. If a node is gone for some time (the implementation currently uses 1 day) then it is garbage collected to bound the size of the recorded metrics. When it re-appears its counter will start from 0. So for the raw metrics in the replicated state it is not always guaranteed that they only go up. We could also offer an API function which hides this detail from the user, but we'd currently prefer not to because it keeps the replica implementation simple and it is consistent with the way Prometheus handles such things.
> 
> Or are you saying that it will never happen that a node disappears for a day?

haha, I guess it's me who doesn't understand how things work :) 
let me ask some questions.

Is it correct that we should have a record for a node for sure if the node is rank 0 at least once per interval, because in this case either `num_blocks_total` or `num_block_failures_total` will be increased in the replicated state based on information passed by consensus? 
This should happen regardless of whether the node in question is actively participating in the system or not.

If the node belongs to the subnet according to the registry versions being used in the subnet for more than 1h, then probability to never have been blockmaker at least one is < (1-1/13)^3600 = 7.184846e-126 in an app subnet and 1.6155048e-20 for the NNS (if I try to do it for 23h hours google tells me it's 0 :)).

If a node belongs to the subnet according to the registry, but is offline for a day, then `num_block_failures_total` will be increasing with very high probability.

> 
> > Last element: To keep the interface cleaner I'd follow Alin's suggestion to
> > have a separate endpoint for this to avoid any confusion and ambiguity.
> 
> No strong opinion. Once there is consensus please ping us so that we can provide an appropriate interface.

Reply:
Yes, I think this is correct. The case described above would only happen if a node is removed and then added back after some time. Admittedly also an unlikely case but we should still handle it properly.

Reply:
Thanks for your explanations, this helped a lot!

In order to do proper accounting, I expect information about all nodes that have been in the subnet for some time in the last 60 days is necessary, isn't it @sasa-tomic?

Therefore I would suggest the following:
1. separate endpoint for current metrics (to keep the interface clean and avoid confusion)
2. report node stats for all nodes that have been part of the subnet in the past 60 days

imho doing resets when node rejoin the subnet makes sense, but I think it's more important that @sasa and his team are able to consume the data in a suitable form, so I defer this question to him :)

Reply:
Makes sense. 

> report node stats for all nodes that have been part of the subnet in the past 60 days

Not sure I understand this. We'd do this in any option. Could you clarify?

Reply:
Thanks for the clarification! For me this would be good enough (but clearly needs to be described in the spec), but let's see what @sasa-tomic says.

Reply:
From my view, if we do as @yvonneanne suggested:
1. separate endpoint for current metrics (to keep the interface clean and avoid confusion)
2. report node stats for all nodes that have been part of the subnet in the past 60 days

we should be good to go. We only need to provide consistency within a 30 day window, where this window is somewhere in the last 60 days, in order to distribute rewards accordingly, and allow the community to verify data, again in this window. After the window is gone, it wouldn't be my biggest problem anymore.

The open question for me are the gaps with missing nodes in the stats within the 60 day window, as @derlerd-dfinity wrote above. I'll sync with him directly to see if there is something easy we can do about this.

Reply:
Had a discussion with @sasa-tomic. We decided that it is the most pragmatic thing to expose the metrics including counter resets. Mainly because the planned client will treat nodes which are missing for an entire day as misbehaving anyways. So one can filter out those first; what remains are only metrics which don't have counter resets. 

Later, if there is demand, we can always add a second API which provides the metrics adjusted for counter resets. 


So for the spec this concretely means the following changes:

- Introduce a separate endpoint for the current metrics.
- Add a note that users of these API have to account for counter resets in the historic metrics, i.e., that if a node doesn't report metrics for a day it will not appear in the metrics for this day and its counters will start from 0 when it re-appears after that.

@sasa-tomic please correct/extend the above in case I misunderstood/missed anything.

Reply:
Yes, that sounds good to me, thanks for writing down the summary, @derlerd-dfinity. Also, as we discussed in person, we don't need to have a definite and final version from the first go. We can make a couple of iterations to come up with a design acceptable to all stakeholders. The key intention here is to 1) appropriately reward well behaving nodes, and 2) penalize misbehaving nodes, both without adding a lot of complexity and strain on our side. So let's go with the cleanest design for now, and we can later discuss where we need to make some changes.

Reply:
Thanks for everyone's comments and especially the summary. What do you think of the following additions to ic.did and this file respectively? 

```
current_node_metrics: (subnet_id : principal;) -> (record {
    timestamp_nanos : nat64;
    node_metrics : vec node_metrics;
  });
```

```
... the number of failed block proposals by this node.

Note that the counters for both the number of blocks proposed and failed block proposals for a node are reset to zero whenever a node is removed and re-added to a subnet's membership.
```

Reply:
Actually, do we even need the `current_node_metrics()` endpoint at this time? who will use it in the near future?

Reply:
I don't see the immediate need for `current_node_metrics()` but maybe someone else does?

Reply:
I'm also fine with not providing `current_node_metrics` this at all. Regarding the note, I'd state this a bit more abstractly. Maybe something along these lines:

```suggestion
Given a subnet ID as input, this method returns a time series of node metrics (field `node_metrics`). The timestamps are represented as nanoseconds since 1970-01-01 (field `timestamp_nanos`) at which the metrics were sampled. The returned timestamps are all timestamps after (and including) the provided timestamp (field `start_at_timestamp_nanos`) for which node metrics are available. The maximum number of returned timestamps is 60 and no two returned timestamps belong to the same UTC day.

Note that a sample will only include metrics for nodes whose metrics changed compared to the previous sample. This means that if a node disappears in one sample and later reappears its metrics will restart from 0 and consumers of this API need to adjust for these resets when aggregating over multiple samples.
```